### DISCUSSION TITLE
Investigate rustaceanvim instead of rust-tools.nvim for Rust extras

### DISCUSSION THREAD

**Discussion by trevarj** [LazyVim/LazyVim#1976] [Ideas]
There hasn't been an update on [rust-tools.nvim ](https://github.com/simrat39/rust-tools.nvim) in a few months, and a new fork has appeared with many modifications called [rustaceanvim](https://github.com/mrcjkb/rustaceanvim). It may be worthwhile to have a look at this fork and see if it fits with Lazy's Rust extras package. 

At a glance, it works pretty much the same and is capable of being a drop-in-replacement. One thing that may be different is [the way it handles setup](https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#quick-setup)

**Comment by neuromaancer** [comment]
I think it would be great if we use [rustaceanvim](https://github.com/mrcjkb/rustaceanvim) instead!

**Reply by neuromaancer** [reply]
Probably I can start a pull request for this.

**Reply by trevarj** [reply]
Have you used it or explored the differences? Was hoping that someone more experienced than myself could dissect it and understand if it's worth switching.

**Reply by neuromaancer** [reply]
I never used it actually, but I think it is pretty promising from its documentation. Well, if someone is capable to do it fast, I would be happy, but no one is doing now, and I think there are some bugs in the original settings (you can find them in the issues and bugs).

**Comment by trevarj** [comment]
After some usage, it is probably not worth migrating until nvim 0.10 releases since `rustaceaonvim` uses the built-in inlay-hints.

Working config, as of today:

lsp.lua
```
  {
    "mrcjkb/rustaceanvim",
    -- version = "^3", -- Recommended
    ft = { "rust" },
    keys = {},
  },
```

options.lua
```
vim.g.rustaceanvim = {
  -- Plugin configuration
  tools = {},
  -- LSP configuration
  server = {
    on_attach = function(client, bufnr)
      vim.lsp.inlay_hint.enable(bufnr)
      -- you can also put keymaps in here
    end,
    settings = {
      -- rust-analyzer language server configuration
      ["rust-analyzer"] = {
        cargo = {
          allFeatures = true,
          loadOutDirsFromCheck = true,
          runBuildScripts = true,
        },
        -- Add clippy lints for Rust.
        checkOnSave = {
          allFeatures = false,
          command = "clippy",
          extraArgs = { "--no-deps" },
        },
        procMacro = {
          enable = true,
          ignored = {
            ["async-trait"] = { "async_trait" },
            ["napi-derive"] = { "napi" },
            ["async-recursion"] = { "async_recursion" },
          },
        },
      },
    },
  },
  -- DAP configuration
  dap = {},
}
```

**Comment by alejandrodnm** [comment]
I tried really quick by copying the config from rust-tools to rustacean as https://github.com/mrcjkb/rustaceanvim/blob/e0853ead50e6327c037a4ab3e3c54fdb140536c2/troubleshooting/minimal.lua#L40-L49

But I had some issues. 

I'm on 0.10-dev.

This was some time ago, I reverted everything.

**Reply by trevarj** [reply]
Did you try my config that I posted above?

**Reply by alejandrodnm** [reply]
No, I didn't try. I found this discussion after trying and reverting everything back.

BTW I got it to work, but got some weird intermittent issues.

I'll try with your config, the thing is that this week is going to be crazy for me. I'll post anything I find.

**Comment by alejandrodnm** [comment]
I've just tried with:

```lua
  {
    "mrcjkb/rustaceanvim",
    version = "^3", -- Recommended
    ft = { "rust" },
    lazy = true,
    init = function()
      vim.g.rustaceanvim = function()
        local ok, mason_registry = pcall(require, "mason-registry")
        local adapter ---@type any
        if ok then
          -- rust tools configuration for debugging support
          local codelldb = mason_registry.get_package("codelldb")
          local extension_path = codelldb:get_install_path() .. "/extension/"
          local codelldb_path = extension_path .. "adapter/codelldb"
          local liblldb_path = ""
          if vim.loop.os_uname().sysname:find("Windows") then
            liblldb_path = extension_path .. "lldb\\bin\\liblldb.dll"
          elseif vim.fn.has("mac") == 1 then
            liblldb_path = extension_path .. "lldb/lib/liblldb.dylib"
          else
            liblldb_path = extension_path .. "lldb/lib/liblldb.so"
          end
          adapter = require("rustaceanvim.config").get_codelldb_adapter(codelldb_path, liblldb_path)
        end
        return {
          dap = adapter,

          server = {
            on_attach = function(_, bufnr)
              vim.lsp.inlay_hint.enable(bufnr, true)
              vim.keymap.set("n", "K", "<cmd>RustLSP hover actions<cr>", { desc = "Hover Actions (Rust)" })
              vim.keymap.set("n", "<leader>cR", "<cmd>RustLsp codeAction<cr>", { desc = "Code Action (Rust)" })
              vim.keymap.set("n", "<leader>dr", "<cmd>RustLsp debuggables<cr>", { desc = "Run Debuggables (Rust)" })
            end,
            settings = {
              -- rust-analyzer language server configuration
              ["rust-analyzer"] = {
                cargo = {
                  allFeatures = true,
                  loadOutDirsFromCheck = true,
                  runBuildScripts = true,
                },
                -- Add clippy lints for Rust.
                checkOnSave = {
                  allFeatures = true,
                  command = "clippy",
                  extraArgs = { "--no-deps" },
                },
                procMacro = {
                  enable = true,
                  ignored = {
                    ["async-trait"] = { "async_trait" },
                    ["napi-derive"] = { "napi" },
                    ["async-recursion"] = { "async_recursion" },
                  },
                },
              },
            },
          },
          tools = {
            on_initialized = function()
              vim.cmd([[
                  augroup RustLSP
                    autocmd CursorHold                      *.rs silent! lua vim.lsp.buf.document_highlight()
                    autocmd CursorMoved,InsertEnter         *.rs silent! lua vim.lsp.buf.clear_references()
                    autocmd BufEnter,CursorHold,InsertLeave *.rs silent! lua vim.lsp.codelens.refresh()
                  augroup END
                ]])
            end,
          },
        }
      end
    end,
  },
```

And it's working. I had to disable some things:

```
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        rust_analyzer = {},
      },
      setup = {
        rust_analyzer = function()
          return true
        end,
      },
    },
},
  {
    "simrat39/rust-tools.nvim",
    enabled = false,
  },
```

Debugging worked. I didn't test extensively.

~Inlay hints are not enabled by default, but calling the Toggle Inlay Hints command (default is <leader>uh) shows them~ I added `vim.lsp.inlay_hint.enable(bufnr, true)` to on_attach to display inlay hints by default. Since this project removed the code for the inlay hints, and it's using neovim's, they are different.

**Reply by trevarj** [reply]
I noticed the inlay hint thing too. Would be good to know if that is a LazyVim thing or rustaceanvim. I haven't tried any other LSP with inlay hints using the built-in neovim ones

**Reply by alejandrodnm** [reply]
I added vim.lsp.inlay_hint.enable(bufnr, true) to on_attach to display inlay hints by default, I updated the comment to reflect this :)

**Reply by dpetka2001** [reply]
There's also #2198. Would be nice if you could check out the pr and try it and maybe make suggestions for improvement if any necessary?

**Comment by trevarj** [comment]
@dpetka2001 thanks. Looks like https://github.com/LazyVim/LazyVim/pull/2198 and https://github.com/LazyVim/LazyVim/issues/2113 are enough to close this discussion and move over to there
